File: app/src/main/java/cn/intret/app/picgo/model/MediaFile.java
package cn.intret.app.picgo.model;

import android.util.Size;

import java.io.File;
import java.util.Comparator;
import java.util.Date;

/**
 * Media file (image/video)
 */
public class MediaFile {

    // TODO: add sort by resolution
    public final static Comparator<MediaFile> MEDIA_FILE_DATE_DESC_COMPARATOR =
            (MediaFile o1, MediaFile o2) -> o2.getDate().compareTo(o1.getDate());
    public final static Comparator<MediaFile> MEDIA_FILE_DATE_ASC_COMPARATOR =
            (MediaFile o1, MediaFile o2) -> o1.getDate().compareTo(o2.getDate());

    public final static Comparator<MediaFile> MEDIA_FILE_NAME_ASC_COMPARATOR =
            (MediaFile o1, MediaFile o2) -> {
                return o1 == null
                        ? (o2 == null ? 0 : 1)
                        : o1.getFile().compareTo(o2.getFile());
            };

    public final static Comparator<MediaFile> MEDIA_FILE_NAME_DESC_COMPARATOR =
            (MediaFile o1, MediaFile o2) -> o2.getFile().compareTo(o1.getFile());


    File mFile;
    Date mDate;

    /**
     * Image size or video resolution
     */
    Size mMediaResolution;

    /**
     * File size ( disk usage of file), in bytes
     */
    long mFileSize;


    long mVideoDuration;

    /**
     * Get duration, in ms
     *
     * @return
     */
    public long getVideoDuration() {
        return mVideoDuration;
    }

    public MediaFile setVideoDuration(long videoDuration) {
        mVideoDuration = videoDuration;
        return this;
    }

    public long getFileSize() {
        return mFileSize;
    }

    public MediaFile setFileSize(long fileSize) {
        mFileSize = fileSize;
        return this;
    }

    public Size getMediaResolution() {
        return mMediaResolution;
    }

    public MediaFile setMediaResolution(Size mediaResolution) {
        mMediaResolution = mediaResolution;
        return this;
    }

    public File getFile() {
        return mFile;
    }

    public void setFile(File file) {
        mFile = file;
    }

    public Date getDate() {
        return mDate;
    }

    public MediaFile setDate(Date date) {
        mDate = date;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof MediaFile)) return false;

        MediaFile mediaFile = (MediaFile) o;

        return mFile != null ? mFile.equals(mediaFile.mFile) : mediaFile.mFile == null;

    }

    @Override
    public int hashCode() {
        return mFile != null ? mFile.hashCode() : 0;
    }

    @Override
    public String toString() {
        return "MediaFile{" +
                "mFile=" + mFile +
                ", mDate=" + mDate +
                ", mMediaResolution=" + mMediaResolution +
                ", mFileSize=" + mFileSize +
                ", mVideoDuration=" + mVideoDuration +
                '}';
    }
}

File: app/src/main/java/cn/intret/app/picgo/model/image/data/MediaFile.kt
package cn.intret.app.picgo.model.image.data

import android.util.Size
import java.io.File
import java.util.*

/**
 * Media file (image/video)
 */
class MediaFile {


    var file: File? = null
    internal lateinit var date: Date

    /**
     * Image size or video resolution
     */
    internal lateinit var mediaResolution: Size

    /**
     * File size ( disk usage of file), in bytes
     */
    internal var fileSize: Long = 0


    internal var videoDuration: Long = 0

    /**
     * Get duration, in ms
     *
     * @return
     */
    fun getVideoDuration(): Long {
        return videoDuration
    }

    fun setVideoDuration(videoDuration: Long): MediaFile {
        this.videoDuration = videoDuration
        return this
    }

    fun getFileSize(): Long {
        return fileSize
    }

    fun setFileSize(fileSize: Long): MediaFile {
        this.fileSize = fileSize
        return this
    }

    fun getMediaResolution(): Size {
        return mediaResolution
    }

    fun setMediaResolution(mediaResolution: Size): MediaFile {
        this.mediaResolution = mediaResolution
        return this
    }

    fun getDate(): Date {
        return date
    }

    fun setDate(date: Date): MediaFile {
        this.date = date
        return this
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is MediaFile) return false

        val mediaFile = o as MediaFile?

        return if (file != null) file == mediaFile!!.file else mediaFile!!.file == null

    }

    override fun hashCode(): Int {
        return if (file != null) file!!.hashCode() else 0
    }

    override fun toString(): String {
        return "MediaFile{" +
                "file=" + file +
                ", date=" + date +
                ", mediaResolution=" + mediaResolution +
                ", fileSize=" + fileSize +
                ", videoDuration=" + videoDuration +
                '}'.toString()
    }

    companion object {

        // TODO: add sort by resolution
        val MEDIA_FILE_DATE_DESC_COMPARATOR = Comparator<MediaFile>{ o1: MediaFile, o2: MediaFile -> o2.getDate().compareTo(o1.getDate()) }
        val MEDIA_FILE_DATE_ASC_COMPARATOR = Comparator<MediaFile>{ o1: MediaFile, o2: MediaFile -> o1.getDate().compareTo(o2.getDate()) }

        val MEDIA_FILE_NAME_ASC_COMPARATOR = Comparator<MediaFile>{ o1: MediaFile, o2: MediaFile -> o1.file!!.compareTo(o2.file) }

        val MEDIA_FILE_NAME_DESC_COMPARATOR = Comparator<MediaFile>{ o1: MediaFile, o2: MediaFile -> o2.file!!.compareTo(o1.file) }

        val MEDIA_FILE_LENGTH_DESC_COMPARATOR = Comparator<MediaFile>{ o1: MediaFile, o2: MediaFile -> java.lang.Long.compare(o2.getFileSize(), o1.getFileSize()) }


        val MEDIA_FILE_LENGTH_ASC_COMPARATOR = Comparator<MediaFile>{ o1: MediaFile, o2: MediaFile -> java.lang.Long.compare(o1.getFileSize(), o2.getFileSize()) }
    }
}
